### ISSUE TITLE
Reinstate syntax highlighting.

### ISSUE THREAD

**Issue by skx** [skx/kilua#51]
There are two approaches here:
- Use a proper parser.
- Clone the old code.

The former has the advantage that it will be easier.  The latter has the advantage Lua can be used.

Either is fine.

**Comment by skx** [comment]
This is interesting reading, using lpeg:
- https://github.com/martanne/vis/issues/81

Is it a dependency too far (paged for Debian :) )?  Will handle #52 before I leave for work, then come back to this in the evening.

**Comment by skx** [comment]
I have a proof of concept for syntax highlighting, which is **entirely** Lua-driven :)

The observation is that we draw our screen thusly:
- Clear the screen
- For each row.
  - For each character - draw the character.

We draw character by character and colours only change on character boundaries.  What does that mean?  It means that we can efficiently store _regions_ of a particular colour:
- character 0-10 in the file are red.
- character 11 is blue.
- character 12-244 is green.

Simples.  We just need to keep track of our current character offset, and set the colour before drawing each one.

That takes care of drawing.  How about setting?  Setting, it turns out is simple.  Consider the attached program

```
  $ lua5.2 c-syntax.lua.txt test.c.txt
   PREPROCESSOR 8   #include
   LITERAL  11  "foo.bar.h"
   PREPROCESSOR 8   #include
   INCLUDE  11  <foo/bar.h>
   COMMENT  15  /**
    * Test
    */
   KEYWORD  3   int
```

What does that tell us?  It tells use that the first token is "PREPROCESSOR" and it is "8 bytes" long.  So we set the colour of the range 0-7 to be "blue".  The next token is a literal string, 11 bytes long, so we set the colour of 8-19 to be "red".  etc.

In short we can parse syntax via `l-peg`, and the parser only has to do two jobs:
- Keep a running total of the lenght of the input parsed thus-far.
- Update ranges, probably via a Lua call `set_range( start, end, "pink" )`

That's actually a hell of a lot easier to debug and extend than the C-based approach that @antirez pioneered, and I don't see any obvious flaws in it.  (True I've not yet implemented it ..)

[c-syntax.lua.txt](https://github.com/skx/kilua/files/389117/c-syntax.lua.txt)
[test.c.txt](https://github.com/skx/kilua/files/389118/test.c.txt)

**Comment by skx** [comment]
So plan:
- Export `set_range_colour` to Lua.
- Update buffer object to have a syntax-type "c", "c++", "lua", "meh".
- Call `update_syntax()` now and again to refresh the view
  - This will invoke lpeg to do the parsing.
  - The parsers will all be the same, and will invoke `set_range_colour()` to update things.
- `draw_screen()` will then use the region/ranges to set the colours.

I have a proof of concept.  It's hacky, but I can clean it up and make it work I think.

@sdegutis might have useful input, but I suspect @kilofork has been left behind due to the changes I've made here.

**Comment by sdegutis** [comment]
Sorry, probably not going to be able to have useful input anymore. Work is more hectic than normal and looks like it will continue to be indefinitely now. Best of luck 👍

**Comment by skx** [comment]
Appreciate the reply regardless, and I will continue to star/watch/follow your work if it ever continues.  You approached things differently to me, but I think there's a lot to be said for that approach too.

More lua in the world is never a bad thing :)

**Comment by sdegutis** [comment]
@skx My plan from the beginning was flawed. Ultimately I wanted to replace Emacs for myself with something lighter weight, with it still being 100% customizable. But then I either have to write all the modules myself that it would need to get me back up to my current development speed, or convince others to join the effort with me. And that's absolutely never going to happen. But that's not necessarily a bad thing. So I just took it as a learning experience and I'm fine with that. I learned a lot more about how to use terminal escape codes, and I learned a pretty darn good syntax highlighting algorithm from @antirez. I do believe that long-term, programmers can do better than Emacs or Vim or Atom or Sublime Text. (No offense, people who wrote those!) But I'm not going to be the person to make that reality come true.

**Comment by skx** [comment]
Current status:
- Added dependency on `lua-lpeg`.
- Each buffer has a syntax-mode associated with it, empty by default, but will contain things like "c", "lua", "makefile".
- If a buffer has a syntax-mode associated with it then `on_syntax_highlight(text)` will be invoked.

The callback function is expected to update the state of the buffer by calling:

```
 syntax_range( start, end, colour )
```

So, for example, if the first five bytes of the file are to be red it will call:

```
  syntax_range(0,5, 2 )
```

(Where "2 == red").

This actually seems to work, and be responsive, but we've got a bug because the colours are shifted around.  Suspect I have an off-by-one somewhere - probably relating to the newline handing.

Once the off-by-one, or two, are found we can just drop in new files `syn/foo.lua` to provide syntax highlighting named `foo`, which will be neat.

TODO:
- The status-bar should draw the syntax-mode, if set
- The off-by-one should be fixed
- We should configure the path for loading syntax files from, at the moment it is just "`./syn`".
  - Should probably include both `/etc/kilua/syntax` + `~/.kilua/syntax/`.

Making progress.

**Comment by skx** [comment]
Syntax highlighting is "complete" now, but suboptimal.  We've supplied syntax highlighting, via `lpeg` for:
- C.
- Lua.
- "markdown".

Scrolling up/down works.  Speed is good.  But there is a problem:
- Syntax highlighting has to be done for the whole buffer.
  - If you scroll down and only attempt to render the text from the top of the screen you might start a line such as "`* This is part of an open comment`".
- So you cannot render line by line, you must take into account the previous history.
  - Though as an optimization you can  truncate the text you render (which we don't) at the point it stops being on-display.
  - i.e. Syntax-highlight (0,0) -> ( `rowoffset + E.screenrows`, 0).
- Because rendering is done buffer-wide, with all the lines, you need to store that globally.  That means the rendering is stored in the buffer - not the `erow` object.
  - I think this is a mistake, because it causes problems when you scroll left/right.

Suggestion:
- Update `erow` to store the arry of characters _and_ the colour details.
- Update the syntax-handling on the lua side to use `RED` not `2` to allow users to change colour-schemes too.
